### PR TITLE
Feature

### PR DESCRIPTION
**Pull request by nx-mukesh**
feat: Intro to Node.js & NPM (#39515) :

### CODE REVIEW

**Comment by nx-mukesh** [comment]
feat: ts configuration (#39520)

**Review by namanpar — CHANGES_REQUESTED** [review]
Kindly remove node_modules files and package-lock.json from pr.
Do commit gitIgnore, package.json and babelrc

**Comment by nx-mukesh** [comment]
feat: babel setup, validation function, helper function (#39517, #3919) :

**Comment by nx-mukesh** [comment]
Update: node_module and package-lock.json deleted

File: extra/index.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+let rowForDiamond = 6;
+let rowForEquilateral = 7;
+
+console.log(`Diamond Shape with ${rowForDiamond} row:- `);
+diamondShape(rowForDiamond);
+
+console.log();
+
+console.log(`Equilateral shape with ${rowForEquilateral} row :- `);
+equilateral(rowForEquilateral);
```
Comment (namanpar) [inline extra/index.js:14]: You already executed these function in pattern/index.js file

File: extra/index.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import { diamondShape, equilateral } from "./pattern/index";
```
Comment (namanpar) [inline extra/index.js]: Export from '/pattern' not '/pattern/index'

File: extra/utils/permission.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    
+//   }
+//   return "Module not found";
+// };
+
+
+
+// console.log(hasPermission(permission.getUsers, "head-trainer", "read"));
+// console.log(hasPermission(permission.getUsers, "trainer", "read"));
+// console.log(hasPermission(permission.getUsers, "trainee", "write"));
```
Comment (namanpar) [inline extra/utils/permission.js]: Finish this task

File: extra/utils/validation.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    } else {
+      invalidUsers.push(userId.traineeEmail);
+    }
+
+    if (reviewerEmail) {
+      validUsers.push(userId.reviewerEmail);
+    } else {
+      invalidUsers.push(userId.reviewerEmail);
+    }
+  });
```
Comment (namanpar) [inline extra/utils/validation.js]: Use destructing

File: extraTs/.gitignore [hunk 5]
```diff
@@ -0,0 +1 @@
+dist
```
Comment (namanpar) [inline extraTs/.gitignore]: .gitIgnore file is created in root directory. Kindly remove it from here. Add dist to gitIgnore

File: extraTs/pattern/diamond.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+          }
+          for(j=i;j<=size-1;j++){
+             process.stdout.write("* ");
+          }
+          console.log();
+       }
+    }
+ }
+
+ createDiamondShape(5);
```
Comment (namanpar) [inline extraTs/pattern/diamond.ts]: Add input type and return type of function

File: extraTs/pattern/equilateral.ts [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+       for(var s=size-1;s>=i;s--){
+          process.stdout.write(" ");
+       }
+       for(var j=1;j<=i;j++){
+          process.stdout.write("* ")
+       }
+       console.log();
+    }
+}
+
```
Comment (namanpar) [inline extraTs/pattern/equilateral.ts]: Add input type and return type of function

**Review by namanpar — CHANGES_REQUESTED** [review]
Kindly do requested changes. Be mindful of committing file.
First complete all task till #39520

File: extra/utils/permission.js (hunk 3 above)
Comment (nx-mukesh) [reply extra/utils/permission.js]: Task completed

File: extra/index.js (hunk 2 above)
Comment (nx-mukesh) [reply extra/index.js]: done

**Comment by nx-mukesh** [comment]
Made changes on feedback

File: extraTs/pattern/equilateral.ts (hunk 7 above)
Comment (nx-mukesh) [reply extraTs/pattern/equilateral.ts]: done

File: extraTs/pattern/diamond.ts (hunk 6 above)
Comment (nx-mukesh) [reply extraTs/pattern/diamond.ts]: done

File: extra/constants.js [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+const permission = {
+  getUsers: {
+    all: ["head-trainer"],
+    read: ["trainee", "trainer"],
+    write: ["trainer"],
+    delete: [],
+  },
+};
+
+module.exports = permission;
```
Comment (namanpar) [inline extra/constants.js:11]: use es6

File: extraTs/utils/helpers.ts [hunk 9]
```diff
@@ -0,0 +1,10 @@
+
+const validateEmail = (userEmail) => {
+    if (userEmail) {
+      const regEx = /^[_A-Za-z0-9-||+]+(||.[_A=Za-z0-9-]+)*@successive.tech$/;
+      return regEx.test(userEmail);
+    }
+  };
+  
+
+  export default validateEmail;
```
Comment (namanpar) [inline extraTs/utils/helpers.ts]: Define input and return type

File: extraTs/pattern/diamond.ts [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+        for(var s=1;s<=i;s++){
+           process.stdout.write(" ");
+        }
+        for(j=i;j<=size-1;j++){
+           process.stdout.write("* ");
+        }
+        console.log();
+     }
+  }
+}
```
Comment (namanpar) [inline extraTs/pattern/diamond.ts]: define input and return type

File: extraTs/constants.ts [hunk 11]
```diff
@@ -0,0 +1,11 @@
+// constant value -
+const permission = {
+  getUsers: {
+    all: ["head-trainer"],
+    read: ["trainee", "trainer"],
+    write: ["trainer"],
+    delete: [],
+  },
+};
+
```
Comment (namanpar) [inline extraTs/constants.ts]: define typrscript

File: extraTs/pattern/equilateral.ts [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+    for (let s = row - 1; s >= i; s--) {
+      process.stdout.write(" ");
+    }
+    for (let j = 1; j <= i; j++) {
+      process.stdout.write("* ");
+    }
+    console.log();
+  }
+};
+
```
Comment (namanpar) [inline extraTs/pattern/equilateral.ts]: define input and return type

File: extraTs/utils/validation.ts [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+};
+
+// list of user's emails
+const users = [
+  {
+    traineeEmail: "trainee1@successive.tech",
+    reviewerEmail: "reviewer1@successive.tech",
+  },
+];
+
```
Comment (namanpar) [inline extraTs/utils/validation.ts]: define input and return type

**Review by namanpar — CHANGES_REQUESTED** [review]
Kindly define interfaces file as per #39520. 
Do review your own work and check it against tasks details.

**Review by namanpar — APPROVED** [review]

**Pull request merged**